### PR TITLE
Scripts/Scarlet Enclave: Grand Theft Palomino

### LINKED ISSUE #20844: [3.3.5] Scripts/ScarletEnclave: Grand Theft Palamino (quest 12680) Deliver Stolen Horse (spell 52264)
**Description:**

During the quest Grand Theft Palamino, after stealing and riding back the horse to Salanar the Horseman, the player uses the "Deliver Stolen Horse" ability to gain the final credit to complete the quest. When this happens the player exits/dismouts the horse in order to turn in the quest.

**Current behaviour:**

Using the spell gives the player credit for the quest, but does not exit the player from the horse. The player cannot turn in the quest until they exit the vehicle.

**Expected behaviour:**

Using the spell gives the player quest credit and exits them from the vehicle so they can turn in the quest.
Here is a reference video to show the intended behavior: https://youtu.be/5wlMkJOekxc?t=159

**Steps to reproduce the problem:**

1. Make a deathknight
2. .quest add 12680
3. .go xyz 2218 -5817 101.54 609
4. ride horse back to Death's Breach
5. approach Salanar the Horseman and use "Deliver Stolen Horse"
 
**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

TrinityCore rev. 3e74ef1528db 2017-11-10 23:45:42 +0100 (3.3.5 branch) (Unix, RelWithDebInfo, Static)

**TDB version:**  TDB_full_world_335.63_2017_04_18

**Operating system:** Fedora 26

### PR DESCRIPTION
**Pull request by Sorikoff**
**Changes proposed:**
Based on #20897 quest [Grand Theft Palomino](https://woehead.way-of-elendil.fr/?quest=12680) implementation.

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #20844

**Tests performed:** Does build, works-in-game

**Known issues and TODO list:**
- [ ] Dark Rider of Acherus ignores walk state when chasing (#20938). This is minor issue, and I think PR can be merged before the actual fix.
- [ ] Creature keep following the leader once charmed (#21731). The same as above.

### CODE REVIEW

File: sql/updates/world/3.3.5/9999_99_99_99_world_335.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+DELETE FROM `spell_script_names` WHERE `spell_id`= 52267;
+DELETE FROM `spell_script_names` WHERE `ScriptName`='spell_deliver_stolen_horse';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+(52264, 'spell_deliver_stolen_horse');
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: missing new line.

Comment (illfated) [reply sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: Yes, just add 1 `<Enter>` at the end of the file to follow TC codestyle.

**Comment by illfated** [comment]
I am glad to see the script become event driven instead of a bunch of loose (and often vulnerable to change) timer calls.

File: sql/updates/world/3.3.5/9999_99_99_99_world_335.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+DELETE FROM `spell_script_names` WHERE `spell_id`= 52267;
```
Comment (jackpoz) [inline sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: is this another unrelated spell than the one in "INSERT INTO" ?

Comment (Sorikoff) [reply sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: Yes, despawn is handled in `spell_deliver_stolen_horse`.

**Comment by burbur** [comment]
Nice one! 
Glad to see you have followed the original work and finished the script with some improvements. Thank you!

I've abandoned the pull request because I could not follow every (right) modification request from trinity staff due to work duties! :)
I apologize with the people who were rewieving it because I've forgot about it after a while.

File: sql/updates/world/3.3.5/9999_99_99_99_world_335.sql (hunk 2 above)
Comment (Keader) [inline sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: Always delete by script name, never with just id.

File: src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp [hunk 3]
```diff
@@ -633,94 +633,64 @@ class npc_death_knight_initiate : public CreatureScript
...
+    npc_dark_rider_of_acherus(Creature* creature) : ScriptedAI(creature)
+    {
+        if (TempSummon* summon = me->ToTempSummon())
+            _horseGUID = summon->GetSummonerGUID();
 
-            void UpdateAI(uint32 diff) override
-            {
-                if (!Intro || !TargetGUID)
-                    return;
+        _events.ScheduleEvent(EVENT_START_MOVING, 1s);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp]: it need be done inside of hooks like: Reset() or JustAppeared(), etc.
Not in constructor.

File: src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp [hunk 4]
```diff
@@ -633,94 +633,64 @@ class npc_death_knight_initiate : public CreatureScript
...
-                        default:
-                            break;
-                    }
-                }
-                else
-                    PhaseTimer -= diff;
-            }
+    void UpdateAI(uint32 diff) override
+    {
+        _events.Update(diff);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp]: _events never reset?

File: src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp [hunk 5]
```diff
@@ -633,94 +633,64 @@ class npc_death_knight_initiate : public CreatureScript
...
-            uint32 PhaseTimer;
-            uint32 Phase;
-            bool Intro;
-            ObjectGuid TargetGUID;
-        };
+    void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+    {
+        if (spell->Id == SPELL_DESPAWN_HORSE && target->GetGUID() == _horseGUID)
+            if (Creature* creature = target->ToCreature())
+                creature->DespawnOrUnsummon(2000);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp]: creature->DespawnOrUnsummon(2s);

File: sql/updates/world/3.3.5/9999_99_99_99_world_335.sql (hunk 2 above)
Comment (illfated) [reply sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: What if deleting the `ScriptName` causes 12 other spells to fail?

Comment (jackpoz) [reply sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: scriptname + id

Comment (Keader) [reply sql/updates/world/3.3.5/9999_99_99_99_world_335.sql]: ⏫

**Comment by jackpoz** [comment]
@Keader any change left to do before merging this ?

**Pull request merged**